src/uu/ls/src/ls.rs

```diff
@@ -2245,52 +2232,91 @@ fn enter_directory(
-                            dired::indent(&mut state.out)?;
-                            let dir_name_size = e.p_buf.to_string_lossy().len();
-                            dired::calculate_subdired(dired, dir_name_size);
-                            // inject dir name
-                            dired::add_dir_name(dired, dir_name_size);
-                        }
+                    #[cfg(target_os = "windows")]
+                    let item_md = FileInformation::from_path(&item.p_buf, item.must_dereference)?;
 
-                        show_dir_name(e, &mut state.out, config)?;
-                        writeln!(state.out)?;
-                        enter_directory(e, rd, config, state, listed_ancestors, dired)?;
-                        listed_ancestors
-                            .remove(&FileInformation::from_path(&e.p_buf, e.must_dereference)?);
-                    } else {
+                    #[cfg(not(target_os = "windows"))]
+                    let item_md = match item.get_metadata(&mut state.out) {
+                        Some(md) => md,
+                        None => &get_metadata_with_deref_opt(&item.p_buf, item.must_dereference)?,
+                    };
+
+                    #[cfg(target_os = "windows")]
+                    if item_md == listed_ancestor_md {
                         state.out.flush()?;
-                        show!(LsError::AlreadyListedError(e.p_buf.clone()));
+                        show!(LsError::AlreadyListedError(item.p_buf.clone()));
+                        continue;
                     }
+
+                    #[cfg(not(target_os = "windows"))]
+                    if item_md.dev() == listed_ancestor_md.dev()
+                        && item_md.ino() == listed_ancestor_md.ino()
+                    {
+                        state.out.flush()?;
+                        show!(LsError::AlreadyListedError(item.p_buf.clone()));
+                        continue;
+                    }
+
+                    // when listing several directories in recursive mode, we show
+                    // "dirname:" at the beginning of the file list
+                    writeln!(state.out)?;
+                    if config.dired {
+                        // We already injected the first dir
+                        // Continue with the others
+                        // 2 = \n + \n
+                        dired.padding = 2;
+                        dired::indent(&mut state.out)?;
+                        let dir_name_size = item.p_buf.to_string_lossy().len();
+                        dired::calculate_subdired(dired, dir_name_size);
+                        // inject dir name
+                        dired::add_dir_name(dired, dir_name_size);
+                    }
+
+                    show_dir_name(&item, &mut state.out, config)?;
+                    writeln!(state.out)?;
+                    let mut res = enter_directory(&item, rd, config, state, dired)?;
+                    queue.append(&mut res);
```

Comment:
Are you using this as a queue or a stack ? because from what I see, you're pushing/popping at the end of the vector, which looks like a stack to me.

I think you may be better going with a VecDeque to make an actual queue